serve/mlc_serve/model/sampler.py

```diff
@@ -0,0 +1,472 @@
+            logit_bias_indices.to(device=dev, non_blocking=True),
+            logit_bias_values.to(device=dev, non_blocking=True),
+            past_output_tokens.to(device=dev, non_blocking=True),
+        )
+
+
+@dataclass
+class SamplingMetadata:
+    # mapping for <batch index, tuple(sampling type, greedy/random sampling index)>
+    index_map: dict[int, Tuple[SamplingType, int]]
+    has_random: bool
+    has_greedy: bool
+    apply_top_p_top_k: bool
+    apply_penalty: bool
+    apply_bias: bool
+    # mapping for <batch index, Tuple(logprob, top-k, top-k index)>
+    logprob_index_map: dict[int, Tuple[bool, int, int]]
+    has_logprob: bool
+    sampling_tensors: SamplingTensors
+    sampling_params: List[SamplingParams]
+
+    @classmethod
+    def from_sampling_params(
+        cls,
+        sampling_params: List[SamplingParams],
+        list_past_output_tokens: List[List[int]],
+        dtype: torch.dtype,
+        dev: str,
+        vocab_size: int,
+    ):
+        list_mask_random = []
+        list_temperatures = []
+        list_top_ps = []
+        list_top_ks = []
+        do_top_p = False
+        do_top_k = False
+        apply_penalty = False
+        apply_bias = False
+        list_frequency_penalties = []
+        list_presence_penalties = []
+        list_repetition_penalties = []
+        list_logit_bias_indices = []
+        list_logit_bias_values = []
+
+        index_map = {}
+        idx_random = -1
+        idx_greedy = -1
+        batch_size = len(sampling_params)
+        # index 0 is for non-logprob requests
+        has_logprob = False
+        logprob_index_map = {}
+        list_mask_top_logprob = np.full(
+            ((LOGPROB_TOP_K_MAX) + 1, batch_size), False, dtype=bool
+        )
+        logit_bias_maxlen = 0
+        idxs_logprob = [-1] * ((LOGPROB_TOP_K_MAX) + 1)
+        for batch_idx, param in enumerate(sampling_params):
+            # Prepare temperature
+            # NOTE: Zero temperature means deterministic sampling
+            # (i.e., greedy sampling or beam search).
+            # Set the temperature to 1 to avoid division by zero.
+            list_temperatures.append(
+                param.temperature if param.temperature >= SAMPLING_EPS else 1.0
+            )
+
+            if param.sampling_type == SamplingType.RANDOM:
+                list_mask_random.append(True)
+                idx_random += 1
+                index_map[batch_idx] = (SamplingType.RANDOM, idx_random)
+                list_top_ps.append(param.top_p)
+                list_top_ks.append(param.top_k if param.top_k != -1 else vocab_size)
+                do_top_p |= list_top_ps[-1] < 1.0 - SAMPLING_EPS
+                do_top_k |= list_top_ks[-1] != vocab_size
+            else:
+                list_mask_random.append(False)
+                idx_greedy += 1
+                index_map[batch_idx] = (SamplingType.GREEDY, idx_greedy)
+
+            # param.top_logprobs is zero if logprob is not used
```

Comment:
Can we improve this comment and encode as assertion or exception.

Reply:
Removed during the code improvement.